internal/controllers/machineset/machineset_preflight.go

```diff
@@ -0,0 +1,211 @@
+				result = util.LowestNonZeroResult(result, checkResult)
+			}
+		}
+
+		// Run the kubeadm-version skew preflight check.
+		if !skipped.Has(clusterv1.MachineSetPreflightCheckKubeadmVersion) {
+			if checkResult, err := r.kubeadmVersionPreflightCheck(ctx, cpSemver, msSemver, ms); err != nil || !checkResult.IsZero() {
+				result = util.LowestNonZeroResult(result, checkResult)
+				errList = append(errList, err)
+			}
+		}
+	}
+
+	return result, errors.Wrap(kerrors.NewAggregate(errList), "failed to perform preflight checks")
+}
+
+func (r *Reconciler) controlPlaneStablePreflightCheck(ctx context.Context, controlPlane *unstructured.Unstructured) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	cpKlogRef := klog.KRef(controlPlane.GetNamespace(), controlPlane.GetName())
+
+	// Check that the control plane is not provisioning.
+	isProvisioning, err := contract.ControlPlane().IsProvisioning(controlPlane)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform %q preflight check: failed to check if ControlPlane %s is provisioning", clusterv1.MachineSetPreflightCheckControlPlaneStable, cpKlogRef)
+	}
+	if isProvisioning {
+		log.Info(fmt.Sprintf("Preflight check %q failed: ControlPlane %s is provisioning", clusterv1.MachineSetPreflightCheckControlPlaneStable, cpKlogRef))
+		return ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}, nil
+	}
+
+	// Check that the control plane is not upgrading.
+	isUpgrading, err := contract.ControlPlane().IsUpgrading(controlPlane)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform %q preflight check: failed to check if the ControlPlane %s is upgrading", clusterv1.MachineSetPreflightCheckControlPlaneStable, cpKlogRef)
+	}
+	if isUpgrading {
+		log.Info(fmt.Sprintf("Preflight check %q failed: ControlPlane %s is upgrading", clusterv1.MachineSetPreflightCheckControlPlaneStable, cpKlogRef))
+		return ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) kubernetesVersionPreflightCheck(ctx context.Context, cpSemver semver.Version, msSemver semver.Version) ctrl.Result {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Check the Kubernetes version skew policy.
+	// => MS minor version cannot be greater than the Control Plane minor version.
+	// => MS minor version cannot be older than 2 minor versions of Control Plane.
+	// Kubernetes skew policy: https://kubernetes.io/releases/version-skew-policy/#kubelet
+	floorVersion := semver.Version{
+		Major: cpSemver.Major,
+		Minor: cpSemver.Minor - 2,
+		Patch: 0,
+	}
+	if msSemver.GT(cpSemver) || msSemver.LT(floorVersion) {
+		log.Info(fmt.Sprintf("Preflight check %q failed: MachineSet version (%s) and ControlPlane version (%s) do not conform to the kubernetes version skew policy", clusterv1.MachineSetPreflightCheckKubernetesVersion, msSemver.String(), cpSemver.String()))
```

Comment:
Let's split those cases in individual ifs with logs which describe the exact error, e.g.:
* Preflight check %q failed: MachineSet version (%s) and ControlPlane version (%s) do not conform to the kubernetes version skew policy: MachineSet version is higher than ControlPlane version
* Preflight check %q failed: MachineSet version (%s) and ControlPlane version (%s) do not conform to the kubernetes version skew policy: MachineSet version is more than 2 minor versions older than the ControlPlane version

Reply:
I think `msSemver.GT(cpSemver)` should also check for the minor version. The patch version doesn't seem to matter: https://kubernetes.io/releases/version-skew-policy/#kubelet